hpx/lcos/gather.hpp

```diff
@@ -11,31 +11,138 @@
 #if defined(DOXYGEN)
 namespace hpx { namespace lcos
 {
+    /// Gather a set of values from different call sites
+    ///
+    /// This function receives a set of values from all call sites operating on
+    /// the given base name.
+    ///
+    /// \param  basename    The base name identifying the gather operation
+    /// \param  result      A future referring to the value to transmit to the
+    ///                     central gather point from this call site.
+    /// \param  generation  The generational counter identifying the sequence
+    ///                     number of the gather operation performed on the
+    ///                     given base name. This is optional and needs to be
+    ///                     supplied only if the gather operation on the given
+    ///                     base name has to be performed more than once.
+    /// \param this_site    The sequence number of this invocation (usually
+    ///                     the locality id). This value is optional and
+    ///                     defaults to whatever hpx::get_ocality_id() returns.
+    ///
+    /// \returns    This function returns a future holding a vector with all
+    ///             gathered values. It will become ready once the gather
+    ///             operation has been completed.
+    ///
     template <typename T>
     hpx::future<std::vector<T> >
     gather_here(char const* basename, hpx::future<T> result,
-        std::size_t num_sites = ~0U, std::size_t this_site = ~0U);
-
+        std::size_t num_sites = std::size_t(-1),
+        std::size_t generation = std::size_t(-1),
```

Comment:
I thought about this. I thought that a user would want to specify a invocation site less frequently as this defaults to the current locality. Thus I moved that one to the end.